js/event.ts

```diff
@@ -0,0 +1,242 @@
+
+  constructor(type: string, {bubbles=false, cancelable=false, composed=false} = {}) {
+    EVENT_ATTRIBUTES.set(this, {
+      type,
+      bubbles,
+      cancelable,
+      composed,
+      currentTarget: null,
+      eventPhase: EventPhase.NONE,
+      isTrusted: false,
+      target: null,
+      timeStamp: Date.now(),
+    });
+  }
+
+  get bubbles(): boolean | TypeError {
+    if (EVENT_ATTRIBUTES.has(this)) {
+      return EVENT_ATTRIBUTES.get(this).bubbles;
+    }
+
+    throw new TypeError('Illegal invocation');
+  }
+
+  get cancelable(): boolean | TypeError {
+    if (EVENT_ATTRIBUTES.has(this)) {
+      return EVENT_ATTRIBUTES.get(this).cancelable;
+    }
+
+    throw new TypeError('Illegal invocation');
+  }
+
+  get composed(): boolean | TypeError {
+    if (EVENT_ATTRIBUTES.has(this)) {
+      return EVENT_ATTRIBUTES.get(this).composed;
+    }
+
+    throw new TypeError('Illegal invocation');
+  }
+
+  get currentTarget(): boolean | TypeError {
+    if (EVENT_ATTRIBUTES.has(this)) {
+      return EVENT_ATTRIBUTES.get(this).currentTarget;
+    }
+
+    throw new TypeError('Illegal invocation');
+  }
+
+  get defaultPrevented(): boolean {
+    return this.canceledFlag;
+  }
+
+  get eventPhase(): boolean | TypeError {
+    if (EVENT_ATTRIBUTES.has(this)) {
+      return EVENT_ATTRIBUTES.get(this).eventPhase;
+    }
+
+    throw new TypeError('Illegal invocation');
+  }
+
+  get isTrusted(): boolean | TypeError {
+    if (EVENT_ATTRIBUTES.has(this)) {
+      return EVENT_ATTRIBUTES.get(this).isTrusted;
+    }
+
+    throw new TypeError('Illegal invocation');
+  }
+
+  get target(): boolean | TypeError {
+    if (EVENT_ATTRIBUTES.has(this)) {
+      return EVENT_ATTRIBUTES.get(this).target;
+    }
+
+    throw new TypeError('Illegal invocation');
+  }
+
+  get timeStamp(): boolean | TypeError {
+    if (EVENT_ATTRIBUTES.has(this)) {
+      return EVENT_ATTRIBUTES.get(this).timeStamp;
+    }
+
+    throw new TypeError('Illegal invocation');
+  }
+
+  get type(): boolean | TypeError {
+    if (EVENT_ATTRIBUTES.has(this)) {
+      return EVENT_ATTRIBUTES.get(this).type;
+    }
+
+    throw new TypeError('Illegal invocation');
+  }
+
+  /** Returns the event’s path (objects on which listeners will be
+   * invoked). This does not include nodes in shadow trees if the
+   * shadow root was created with its ShadowRoot.mode closed.
+   *
+   *      event.composedPath();
+   */
+  composedPath(): domTypes.EventTarget[] {
+    const composedPath = [];
+
+    if (this.path.length === 0) {
+      return composedPath;
+    }
+
+    composedPath.push(this.currentTarget);
+
+    let currentTargetIndex = 0;
+    let currentTargetHiddenSubtreeLevel = 0;
+
+    for (let index = this.path.length - 1; index >= 0; index--) {
+      const { item, rootOfClosedTree, slotInClosedTree } = this.path[index];
+
+      if (rootOfClosedTree) {
+        currentTargetHiddenSubtreeLevel++;
+      }
+
+      if (item === this.currentTarget) {
```

Comment:
Note for @ztplz that the `item` here is some sort of `EventTarget` object, so you will have to create some sort of method or [wrapper](https://github.com/jsdom/jsdom/blob/master/lib/jsdom/living/events/Event-impl.js#L113) to do equality comparisons (same goes for lines 162 and 185).